core/src/test/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java

```diff
@@ -1340,8 +1340,240 @@ public String executor() {
+                    service.sendRequest(node, "action1", new TestRequest("secondary " + request.info),
+                        TransportRequestOptions.builder().withCompress(randomBoolean()).build(),
+                        new TransportResponseHandler<TestResponse>() {
+                            @Override
+                            public TestResponse newInstance() {
+                                return new TestResponse();
+                            }
+
+                            @Override
+                            public void handleResponse(TestResponse response) {
+                                try {
+                                    if (randomBoolean()) {
+                                        Thread.sleep(randomIntBetween(10, 50));
+                                    }
+                                    logger.debug("send secondary response {}", response.info);
+
+                                    channel.sendResponse(response);
+                                } catch (Exception e) {
+                                    throw new RuntimeException(e);
+                                }
+                            }
+
+                            @Override
+                            public void handleException(TransportException exp) {
+                                try {
+                                    logger.debug("send secondary exception response for request {}", request.info);
+                                    channel.sendResponse(exp);
+                                } catch (Exception e) {
+                                    throw new RuntimeException(e);
+                                }
+                            }
+
+                            @Override
+                            public String executor() {
+                                return randomBoolean() ? ThreadPool.Names.SAME : ThreadPool.Names.GENERIC;
+                            }
+                        });
+                } else {
+                    logger.debug("send response for {}", request.info);
+                    channel.sendResponse(new TestResponse("Response for: " + request.info));
+                }
+
+            }
+        }
+        serviceB.registerRequestHandler("action1", TestRequest::new, randomFrom(ThreadPool.Names.SAME, ThreadPool.Names.GENERIC),
+            new TRH(serviceB));
+        serviceC.registerRequestHandler("action1", TestRequest::new, randomFrom(ThreadPool.Names.SAME, ThreadPool.Names.GENERIC),
+            new TRH(serviceC));
+        serviceA.registerRequestHandler("action1", TestRequest::new, randomFrom(ThreadPool.Names.SAME, ThreadPool.Names.GENERIC),
+            new TRH(serviceA));
+        int iters = randomIntBetween(30, 60);
+        CountDownLatch allRequestsDone = new CountDownLatch(iters);
+        class TRH2 implements TransportResponseHandler<TestResponse> {
```

Comment:
Nit: Can we give this a more meaningful name instead of an abbreviation? I'm with `TestTransportResponseHandler` for example.